### PR TITLE
Gateway requests with federation id

### PR DESCRIPTION
**Pull request by okjodom**
Adds a `federation_id` field in Gateway requests so the Gateway knows what federation to serve.
This is a prerequisite to #664

### CODE REVIEW

File: gateway/ln-gateway/src/ln.rs
Comment (justinmoon) [inline gateway/ln-gateway/src/ln.rs:69]: These structs are specific to CLN so they should probably stay in `cln.rs`

Comment (okjodom) [reply gateway/ln-gateway/src/ln.rs:69]: Ack. I'm hoping LND helps us define generic versions of these so the signature of `ReceiveInvoice` more or less stays the same

File: gateway/ln-gateway/src/lib.rs
Comment (justinmoon) [inline gateway/ln-gateway/src/lib.rs:84]: I don't think this is a good name. Maybe `IncomingHtlc` or something. There's no invoice involved here.

File: gateway/ln-gateway/src/webserver.rs
Comment (justinmoon) [inline gateway/ln-gateway/src/webserver.rs:102]: I think before we merge this we need to nail down what a "federation id" is. Using the name here isn't a good idea because presumably that could change.

File: gateway/ln-gateway/src/ln.rs
Comment (justinmoon) [reply gateway/ln-gateway/src/ln.rs]: I expect we'll have 2 sets of structs that are different. These are CLN / LND implementation details.

Once we implement LND we can make our own struct that just contains the information which the gateway cares about. We could implement `Into` for between the CLN / LND specific structs into the version that the gateway handles.

File: gateway/ln-gateway/src/lib.rs
Comment (okjodom) [reply gateway/ln-gateway/src/lib.rs:84]: `HtlcAccepted` translated to me like, "gateway receive invoice on behalf of user". I propose "ReceiveInvoice" on this basis, given the opposite direction is named "PayInvoice" which is nicely transparent.

Comment (okjodom) [reply gateway/ln-gateway/src/lib.rs:84]: `HtlcAccepted` translated to me like, "gateway receive invoice on behalf of user". I propose "ReceiveInvoice" on this basis, given the opposite direction is named "PayInvoice" which is nicely transparent.

Comment (okjodom) [reply gateway/ln-gateway/src/lib.rs:84]: `HtlcAccepted` translated to me like, "gateway receive invoice on behalf of user". I propose "ReceiveInvoice" on this basis, given the opposite direction is named "PayInvoice" which is nicely transparent.

File: gateway/ln-gateway/src/webserver.rs
Comment (okjodom) [reply gateway/ln-gateway/src/webserver.rs:102]: Yeap. Makes sense to further that discussion

Comment (okjodom) [reply gateway/ln-gateway/src/webserver.rs:102]: This was my attempt to plan for any change in the value we end up using as Federation ID. I defined the ID as a struct in gateway so changes to the value does not result in any churn in gateway code:
```
// Placeholder struct for identifying federations within a gateway
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct FederationId(pub String);
```
IMO the signature should remain rather stable regardless of what we adopt. I didn't want to be super blocked by the discussion and implementation of the ID in Federation

Comment (okjodom) [reply gateway/ln-gateway/src/webserver.rs:102]: This was my attempt to plan for any change in the value we end up using as Federation ID. I defined the ID as a struct in gateway so changes to the value does not result in any churn in gateway code:
```
// Placeholder struct for identifying federations within a gateway
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct FederationId(pub String);
```
IMO the signature should remain rather stable regardless of what we adopt. I didn't want to be super blocked by the discussion and implementation of the ID in Federation

File: gateway/cli/src/main.rs
Comment (elsirion) [inline gateway/cli/src/main.rs:28]: Can we use the newtype instead here? I think it either only needs an annotation or an impl of `FromStr` to work.

File: gateway/ln-gateway/src/lib.rs
Comment (elsirion) [reply gateway/ln-gateway/src/lib.rs:84]: For the purposes of this PR the federation id can be user defined afaik? You could just give it any id on join.

While I do have a slight preference for making this id deterministically derived from the config and using this right away as `FederationId(Sha256)` instead of `FederationId(String)` I'd be ok with this for now if it unblocks you and accelerates resolution of #807 which is the highest priority on the GW rn imo.

**Review by elsirion — COMMENTED** [review]
I don't really see the code do anything with the id? Could that be added in the same PR? As it is it seems to add a bunch of dead code.

File: gateway/ln-gateway/src/webserver.rs
Comment (elsirion) [reply gateway/ln-gateway/src/webserver.rs:102]: For the purposes of this PR the federation id can be user defined afaik? You could just give it any id on join.

While I do have a slight preference for making this id deterministically derived from the config and using this right away as `FederationId(Sha256)` instead of `FederationId(String)` I'd be ok with this for now if it unblocks you and accelerates resolution of #807 which is the highest priority on the GW rn imo.

**Comment by okjodom** [comment]
> I don't really see the code do anything with the id? Could that be added in the same PR? As it is it seems to add a bunch of dead code.

@elsirion , #794 is where I start using federation ids. The code is not dead for long. Though, I have to admit I have a number of works in progress here it's not that obvious how they all tie together

Now, I have to rework 794 given early reviews by Justin, and my own observation of complexity of my current approach.
The federation ID aspect of that change is one stable part I think I can frontload and check in as proposed by this PR :)

File: gateway/ln-gateway/src/webserver.rs
Comment (okjodom) [reply gateway/ln-gateway/src/webserver.rs:102]: Now switching fully to the high pri auth work. It's not related / blocked by this PR at all. Still, I wanted to leave this PR in a merge-able state so I don't have too many hanging conversations

**Comment by elsirion** [comment]
Sounds good, do you know if the [current CI failure](https://github.com/fedimint/fedimint/actions/runs/3305068342/attempts/1) is related to the PR or just general flakyness?

**Comment by okjodom** [comment]
> Sounds good, do you know if the [current CI failure](https://github.com/fedimint/fedimint/actions/runs/3305068342/attempts/1) is related to the PR or just general flakyness? 

A job rerun might help me confirm. The latency test failure looks different from what I investigated on flaky code coverage and build CIs

**Comment by elsirion** [comment]
Re-run fixed it, but GH complains that the branch is out of date, could you rebase on master please?

**Review by elsirion — APPROVED** [review]

**Review by elsirion — APPROVED** [review]

**Pull request merged**
